RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc

```diff
@@ -22,346 +25,351 @@ void HGCalCLUEAlgo::populate(const HGCRecHitCollection &hits) {
-  // if it is zero we are  done
-  if (nClustersOnLayer == 0) return nClustersOnLayer;
-
-  // assign remaining points to clusters, using the nearestHigher set from
-  // previous step (always set except
-  // for top density hit that is skipped...)
-  for (unsigned int oi = 1; oi < nd_size; ++oi) {
-    unsigned int i = rs[oi];
-    int ci = nd[i].data.clusterIndex;
-    if (ci == -1 && nd[i].data.delta < 2. * delta_c) {
-      nd[i].data.clusterIndex = nd[nd[i].data.nearestHigher].data.clusterIndex;
+    bool foundNearestHigherInSearchBox = (i_delta != maxDelta);
+    if (foundNearestHigherInSearchBox){
+      cellsOnLayer.delta[i] = i_delta;
+      cellsOnLayer.nearestHigher[i] = i_nearestHigher;
+    } else {
+      // otherwise delta is guaranteed to be larger outlierDeltaFactor_*delta_c
+      // we can safely maximize delta to be maxDelta
+      cellsOnLayer.delta[i] = maxDelta;
+      cellsOnLayer.nearestHigher[i] = -1;
     }
   }
+}
+
 
-  // make room in the temporary cluster vector for the additional clusterIndex
-  // clusters
-  // from this layer
-  if (verbosity_ < pINFO) {
-    LogDebug("HGCalCLUEAlgo")
-      << "resizing cluster vector by " << nClustersOnLayer << std::endl;
+int HGCalCLUEAlgo::findAndAssignClusters(const unsigned int layerId, float delta_c ) {
+   
+  // this is called once per layer and endcap...
+  // so when filling the cluster temporary vector of Hexels we resize each time
+  // by the number  of clusters found. This is always equal to the number of
+  // cluster centers...
+  unsigned int nClustersOnLayer = 0;
+  auto& cellsOnLayer = cells_[layerId];
+  unsigned int numberOfCells = cellsOnLayer.detid.size();
+  std::vector<int> localStack;
+  // find cluster seeds and outlier  
+  for(unsigned int i = 0; i < numberOfCells; i++) {
+    float rho_c = kappa_ * cellsOnLayer.sigmaNoise[i];
+    // initialize clusterIndex
+    cellsOnLayer.clusterIndex[i] = -1;
+    bool isSeed = (cellsOnLayer.delta[i] > delta_c) && (cellsOnLayer.rho[i] >= rho_c);
+    bool isOutlier = (cellsOnLayer.delta[i] > outlierDeltaFactor_*delta_c) && (cellsOnLayer.rho[i] < rho_c);
+    if (isSeed) 
+    {
+      cellsOnLayer.clusterIndex[i] = nClustersOnLayer;
+      cellsOnLayer.isSeed[i] = true;
+      nClustersOnLayer++;
+      localStack.push_back(i);
+    
+    } else if (!isOutlier) {
+      cellsOnLayer.followers[cellsOnLayer.nearestHigher[i]].push_back(i);   
+    } 
   }
-  clustersOnLayer.resize(nClustersOnLayer);
-
-  // Fill the cluster vector
-  for (unsigned int i = 0; i < nd_size; ++i) {
-    int ci = nd[i].data.clusterIndex;
-    if (ci != -1) {
-      clustersOnLayer[ci].push_back(nd[i]);
-      if (verbosity_ < pINFO) {
-        LogDebug("HGCalCLUEAlgo")
-          << "Pushing hit " << i << " into cluster with index " << ci << std::endl;
-      }
+  // need to pass clusterIndex to their followers
+  while (!localStack.empty()) {
+    int endStack = localStack.back();
+    // RecHitGPU thisHit = hits[frontOfBuffer];
```

Comment:
delete commented-out code (or add a comment about why it is kept)

Reply:
removed